rfcs/0002-tracing-cache.md

```diff
@@ -0,0 +1,95 @@
+  <dd><a href="URL">URL</a></dd>
+
+  <dt>Obsoletes (if applicable)</dt>
+  <dd>None</dd>
+
+  <dt>Date of submission</dt>
+  <dd>2019-12-13</dd>
+
+  <dt>Date of approval</dt>
+  <dd>YYYY-MM-DD</dd>
+
+  <dt>Approved by</dt>
+  <dd>First Person, Second Person</dd>
+</dl>
+
+## Summary
+
+This RFC proposes the creation of a local etherium tracing cache to speed up indexing of subgraphs which use callHandler.
+
+## Motivation
+
+If a subgraph has callHandlers specified, it is necessary to extract calls from a trace of each block that a graph node indexes. It is expensive to acquire and process traces from etherium nodes, both in money and time.
+
+A production graph node may index multiple subgraphs and repeatedly access the same traces many times - especially for recent blocks. Having that data on hand in a pre-processed and easily filtered format may substantially reduce the cost of acquiring and converting trace data into triggers relevant to a subgraph.
+
+## Urgency
+
+None
+
+## Terminology
+
+None.
+
+## Detailed Design
+
+The design comes in 2 parts. The caching hierarchy, and the easily filtered data format.
+
+### Cache Hierarchy
+In `etherium_adapter.rs` is a function `traces`, which unconditionally makes a call to eth.web3. A hierarchical cache would be inserted here with 3 layers - In-Memory, On-Disk, and Fetch.
+
+#### In-Memory Layer
+At the top level of the cache is the in-memory cache. For a graph node which is indexing multiple subgraphs in parallel, it may be likely that they require the latest blocks simultaneously. By storing the buffer containing immutable trace data for a block in a rc::Weak, a low-risk, low-cost cache is introduced. The most recent trace, and all currently in-use traces would be maintained.
+
+#### On-Disk Layer
+The In-Memory layer is not expected to remain in cache for very long in an effort to reduce memory consumption. An on-disk layer helps to prevent more calls to the fetch layer than is absolutely necessary.
+
+This on-disk layer would be implemented as a binary file store with a cache eviction policy over postgres. It is expected that the most recent blocks would be indexed the most frequently, so the scoring mechanism for the cache would probably be some mix between LRU, block recency, and block finality.
+
+#### Fetch layer
+The existing code in the `traces` function would be renamed to `fetch_traces` and would be the bottom layer of the cache. At first, this would exist like it does today. In the future this may be augmented to either delegate to fetching from a parity API or to calculate traces on demand using a local etherium node, depending on which resources are currently the bottleneck.
+
+#### Futures re-use
+In order to prevent concurrent loads of data from one cache layer to another (Eg: multiple concurrent calls to the underlying eth.web3 API with the same parameters) the hierarchical cache would have a mechanism to delegate multiple tasks to the same future with a NonReady state.
```

Comment:
I'm not sure I get this part; it feels like an implementation detail as well.

Reply:
This is trying to state that when multiple cache misses occur the work should not be performed twice concurrently. It is an implementation detail though, and I'll move it to the engineering plan.